### PR TITLE
Add experimental support for MSC3202: allowing application services to masquerade as specific devices.

### PR DESCRIPTION
**Pull request by reivilibre**
- [x] add tests

somewhat commit-by-commit friendly?

### CODE REVIEW

File: synapse/storage/databases/main/devices.py
Comment (anoadragon453) [inline synapse/storage/databases/main/devices.py:125]: I think it'd be nicer if we just switched `get_device` to return an `Optional`, and then change up the callers to look for `None` instead of a `StoreError`.

That would also fix up callers that seem to currently 500 if a device isn't found(!)

File: tests/api/test_auth.py
Comment (anoadragon453) [inline tests/api/test_auth.py:264]: This one needs to be falsey.

File: tests/api/test_auth.py
Comment (anoadragon453) [inline tests/api/test_auth.py:266]: Hmm, I see the argument for consistency, but does `args={}` actually do anything? Likewise above.

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
Some minor things. Looks good to me otherwise! Was a pleasure to review commit-by-commit.

Handy link to [the MSC](https://github.com/matrix-org/matrix-doc/pull/3202).

File: synapse/storage/databases/main/devices.py
Comment (reivilibre) [reply synapse/storage/databases/main/devices.py:125]: Span out to #11565 .

File: tests/api/test_auth.py
Comment (reivilibre) [reply tests/api/test_auth.py:266]: `args` is where the access token, desired user ID and desired device ID go. Did I misunderstand? Is there a nicer way to do this?

Comment (anoadragon453) [reply tests/api/test_auth.py:266]: Ohhh, hah. I thought `args` was some kind of option that Mock accepted, rather than just laying out the spec for `SynapseRequest`. That makes much more sense.

**Review by anoadragon453 — APPROVED** [review]
lgtm after merging in develop!

File: synapse/api/auth.py
Comment (turt2live) [inline synapse/api/auth.py:177]: for some reason this isn't enough to cause `/account/whoami` to return the device ID back.

Comment (turt2live) [reply synapse/api/auth.py:177]: nevermind, I was given the wrong experimental flag.

**Comment by turt2live** [comment]
Just wanted to drop by and say this works :)

**Pull request merged**
